src/routes/file_utils.py

```diff
@@ -0,0 +1,134 @@
+import abc
+import hashlib
+import logging
+import os
+import pathlib
+import shutil
+from io import BytesIO
+
+from typing import Tuple, Optional
+from typing.io import IO
+
+import boto3
+from urllib import request
+
+from boto3_type_annotations.s3.client import Client as S3Client
+from dotenv import load_dotenv
+
+logger = logging.getLogger(__name__)
+load_dotenv()
+
+LOCAL_FILES_DIRECTORY = os.environ.get('LOCAL_FILES_DIRECTORY') or '/tmp/scihive-papers'
+EXTERNAL_BASE_ADDRESS = os.environ.get('EXTERNAL_BASE_ADDRESS') or 'http://localhost:5000'
+
+s3_key = os.environ.get('S3_KEY')
+s3_secret = os.environ.get('S3_SECRET')
+s3_bucket = os.environ.get('S3_BUCKET_NAME')
+
+s3_available = s3_key and s3_secret and s3_bucket
+s3_client: Optional[S3Client] = boto3.client(
+    's3',
+    aws_access_key_id=s3_key,
+    aws_secret_access_key=s3_secret
+) if s3_available else None
+
+
+class FileAccessProvider(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def exists(self, path: str) -> bool:
+        pass
+
+    @abc.abstractmethod
+    def get_link_to_file(self, path: str) -> str:
+        pass
+
+    @abc.abstractmethod
+    def save_file(self, filename: str, content: IO) -> None:
+        pass
+
+
+class S3FileAccessProvider(FileAccessProvider):
+
+    def __init__(self, s3_client: S3Client, s3_bucket: str, prefix: str) -> None:
+        self._s3_client = s3_client
+        self._s3_bucket = s3_bucket
+        self._prefix = prefix
+
+    def exists(self, path: str) -> bool:
+        prefixed_key = f'{self._prefix}/{path}'
+        response = self._s3_client.list_objects_v2(
+            Bucket=self._s3_bucket,
+            Prefix=prefixed_key,
+        )
+        for obj in response.get('Contents', []):
+            if obj['Key'] == prefixed_key:
+                return True
+        return False
+
+    def get_link_to_file(self, path: str) -> str:
+        return f'https://{self._s3_bucket}.s3.amazonaws.com/{self._prefix}/{path}'
```

Comment:
Can we add another env variable for `s3_public_bucket` to support other DNS records?
We can also fall back to `{self._s3_bucket}.s3.amazonaws.com` if `s3_public_bucket` is missing

Reply:
Sure, though I'm wondering - since it means you'll have to modify env var in production to configure the new one, wouldn't it be easier to just use the actual bucket name for the BUCKET env var and alleviate the need for the DNS? or are there other reasons for the DNS?

Reply:
Accessing to s3 directly is not ideal because afaik it has no CDN or caching. It's probably negligible though in our case :)
We can also replace the address on the fly when a user asks for a paper. Let's make a decision and move forward. What do you prefer?

Reply:
Ohhhh now I realize how you've been using the DNS :) Without seeing the configuration, it was a little hard to figure out, I thought you just used the DNS as a redirect to the bucket name. Yeah, direct links to S3 aren't great, DNS is much better though I would recommend separating the bucket name in the config from the DNS prefix. I'll do that, let me know how that works.

Reply:
> Can we add another env variable for s3_public_bucket to support other DNS records?
We can also fall back to {self._s3_bucket}.s3.amazonaws.com if s3_public_bucket is missing

Like this?